### DISCUSSION TITLE
Do I need to define build envs in Vercel?

### DISCUSSION THREAD

**Discussion by willneill** [BetaHuhn/deploy-to-vercel-action#149] [Q&A]
Perhaps this is a noob question, but when passing environment variables to the build as [described here](https://github.com/BetaHuhn/deploy-to-vercel-action#pass-environment-variables-to-the-build) it's unclear whether these env vars also need to be defined in Vercel.

```
BUILD_ENV: |
  FOO="bar"
  SOME_TOKEN="${{ secrets.SOME_TOKEN }}"
```

Because this is a GitHub action, I'd assumed this meant `SOME_TOKEN="${{ secrets.SOME_TOKEN }}"` would include a GitHub secret in the payload to Vercel. Meaning env variables and secrets could be defined in GitHub actions rather than Vercel.

I'm finding that this isn't the case though. It seems that env vars need to be created in Vercel, and that this part of the action is more like a pointer to tell Vercel which env vars you want a specific environment to use.

Some more clarity about this would be really helpful. Thanks

**Comment by BetaHuhn** [comment]
The action uses the Vercel CLI under the hood and only passes the provided variables to it, see #144.

Variables defined with `BUILD_ENV` are only used during the Vercel build step and are not available for any Vercel deployments. For more information check out the [Vercel docs](https://vercel.com/docs/cli#introduction/unique-options/build-env).

Hope this clears things up! Let me know if you have any more questions.

**Reply by willneill** [reply]
Hey @BetaHuhn, yes that clears up a lot actually. Thanks for the super quick reply!

**Reply by BetaHuhn** [reply]
Great! No problem, happy to help 😄

**Reply by willneill** [reply]
Hi @BetaHuhn, I'm still facing some challenges with the `BUILD_ENV`.

If I run the following in a GitHub workflow I successfully start a build in Vercel, but the `BUILD_ENV` vars (specifically the Firebase ones) seem to be missing in the Vercel build step.

```
... other build steps

# deploy to Vercel
- name: Deploy to Vercel
  uses: BetaHuhn/deploy-to-vercel-action@v1
  with:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
    VERCEL_TOKEN: ${{ secrets.VERCEL_TOKEN }}
    VERCEL_ORG_ID: ${{ secrets.VERCEL_ORG_ID }}
    VERCEL_PROJECT_ID: ${{ secrets.VERCEL_PROJECT_ID }}
    BUILD_ENV: | 
      NEXT_PUBLIC_BUILD_VERSION="${{ needs.setup.outputs.version }}"
      NEXT_PUBLIC_BUILD_ENV="${{ needs.setup.outputs.branch }}"
      NEXT_PUBLIC_API_KEY="${{ secrets.FIREBASE_API_KEY_DEVELOPMENT }}"
      NEXT_PUBLIC_AUTH_DOMAIN="${{ secrets.FIREBASE_AUTH_DOMAIN_DEVELOPMENT }}"
      NEXT_PUBLIC_PROJECT_ID="${{ secrets.FIREBASE_PROJECT_ID_DEVELOPMENT }}"
      NEXT_PUBLIC_STORAGE_BUCKET="${{ secrets.FIREBASE_STORAGE_BUCKET_DEVELOPMENT }}"
      NEXT_PUBLIC_MESSAGING_SENDER_ID="${{ secrets.FIREBASE_MESSAGING_SENDER_ID_DEVELOPMENT }}"
      NEXT_PUBLIC_APP_ID="${{ secrets.FIREBASE_APP_ID_DEVELOPMENT }}"
      NEXT_PUBLIC_MEASUREMENT_ID="${{ secrets.FIREBASE_MEASUREMENT_ID_DEVELOPMENT }}"
```

I've tied various approaches to setting the value of the variables, including hardcoded tests. Unfortunately, I haven't been able to send the vars to Vercel from the GitHub deploy action yet.

However, if I use the Vercel CLI locally with something like below, the build works successfully with the correct vars in the build step.

```
vercel --build-env NEXT_PUBLIC_BUILD_VERSION="0.0.1" --build-env NEXT_PUBLIC_BUILD_ENV="development" ... etc 
```

I noticed that there's an [open PR to upgrade the Vercel](https://github.com/BetaHuhn/deploy-to-vercel-action/pull/143). Wondered if this may be a potential cause for this issue?

Appreciate any insight you can share.

**Reply by willneill** [reply]
Looking at this line [here](https://github.com/BetaHuhn/deploy-to-vercel-action/blob/8b684b38089cc692f0831f499da7be73ee2f8e76/src/vercel.js#L58)...

```javascript
if (BUILD_ENV) {
  BUILD_ENV.forEach((item) => {
    commandArguments = commandArguments.concat([ '--build-env', item ])
  })
}
```

Should this perhaps have a space after env like this '--build-env ' ?
See [Vercel CLI docs](https://vercel.com/docs/cli#introduction/unique-options/build-env)

**Reply by BetaHuhn** [reply]
> I noticed that there's an https://github.com/BetaHuhn/deploy-to-vercel-action/pull/143. Wondered if this may be a potential cause for this issue?

That shouldn't affect this, looking through the Vercel changelog nothing related to build envs has changed.

> Should this perhaps have a space after env like this '--build-env ' ?

The array is passed to the [child_process](https://nodejs.org/api/child_process.html) `spawn` method which handles the parsing of arguments. It works fine with the `--meta` arguments.

I tested the `BUILD_ENV` input and it seems to work fine for me. Each variable defined with it is present as an environment variable during the build step.

Could you enable the [action debug mode](https://docs.github.com/en/github-ae@latest/actions/monitoring-and-troubleshooting-workflows/enabling-debug-logging#enabling-step-debug-logging) by setting the `ACTIONS_STEP_DEBUG` secret to true and share the logs (excluding sensitive values) @WillNeill? This might help me debug things.
